src/frontends/lean/decl_util.cpp

```diff
@@ -79,12 +141,23 @@ expr parse_single_header(parser & p, declaration_name_scope & scope,
         while (is_pi(it)) it = binding_body(it);
         expr const & C = get_app_fn(it);
         name ns = get_namespace(p.env());
-        if (is_constant(C) && !ns.is_anonymous()) {
-            c_name = const_name(C);
-            scope.set_name(c_name);
-        } else if (is_constant(C) && is_app(it) && is_constant(get_app_fn(app_arg(it)))) {
-            c_name = const_name(get_app_fn(app_arg(it))) + const_name(C);
+        if (is_constant(C) && is_app(it) && is_constant(get_app_fn(app_arg(it)))) {
```

Comment:
Since this is duplicated in two places, I'd factor out the code into a new function:
```c++
optional<name> heuristic_inst_name(expr const & e) {
  // ...
}

  if (auto c_name = heuristic_inst_name(it)) {
    scope.set_name(c_name->replace_prefix(ns, name()));
  }
```

Related to this, I don't think the code should replace the namespace of the argument because this makes it behave differently whether we're in a namespace or not (which is what this PR is supposed to avoid).
```lean
namespace a

def b := unit
class c (α : Type).

-- atm gets name a.c.b
instance : c b := ⟨⟩

end a

-- atm gets name a.c.a.b
instance : a.c a.b := ⟨⟩
```